### PR TITLE
Fix client cache

### PR DESCRIPTION
**Pull request by deads2k**
Fixes https://github.com/openshift/origin/issues/5820.

When we negotiate for "", the "" value is filled in the clientCache, but we don't also fill the "v1" slot.  The `resource.Builder` then requests a RESTClient for its resource and the cache gets filled again.  This fills in the correct slot.

### CODE REVIEW

**Comment by liggitt** [comment]
add a test to `TestClientConfigForVersion` to cover this case?

**Comment by liggitt** [comment]
also, I think the negotiate process should remember the server versions after fetching them the first time... if I already was told the server supports `v1, v2, v3`, I shouldn't need to ask it three times to get three clients for those three versions... follow-up issue

**Comment by deads2k** [comment]
> also, I think the negotiate process should remember the server versions after fetching them the first time... if I already was told the server supports v1, v2, v3, I shouldn't need to ask it three times to get three clients for those three versions... follow-up issue

I'd rather wait until after discovery is better hammered out upstream.  It doesn't make sense to retrieve at all if we cache a copy locally, which makes a lot of sense.  Then expire the local cache on a request for a group we don't have or an API server rejection that indicates unsupported version.

File: pkg/cmd/util/clientcmd/factory.go [hunk 1]
```diff
@@ -489,6 +489,11 @@ func (c *clientCache) ClientConfigForVersion(version string) (*kclient.Config, e
 	client.SetOpenShiftDefaults(&config)
 	c.configs[version] = &config
 
+	// `version` does not necessarily equal `config.Version`.  However, we know that we call this method again with
+	// `config.Version`, we should get the the config we've just built.
+	config = config
```
Comment (deads2k) [inline pkg/cmd/util/clientcmd/factory.go]: @liggitt is there an easy way to make verify happy?

```
pkg/cmd/util/clientcmd/factory.go:494: self-assignment of config to config
```

My intent here is to create another copy to ensure distinct references in the cache.

Comment (deads2k) [inline pkg/cmd/util/clientcmd/factory.go]: > @liggitt is there an easy way to make verify happy?

@stevekuznetsov same question

Comment (liggitt) [inline pkg/cmd/util/clientcmd/factory.go]:
```
configCopy := config
c.configs[config.Version] = &configCopy
```
?

**Comment by deads2k** [comment]
test added.

**Comment by deads2k** [comment]
[test]

**Comment by liggitt** [comment]
[merge]

**Comment by deads2k** [comment]
UI flake @jwforres do you have one to cover this?  re[merge]

```

**************************************************
*                    Failures                    *
**************************************************

1)  authenticated e2e-user new project when creating a new project should validate taken name when trying to create
  - Error: Wait timed out after 3015ms
      at /data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver/promise.js:1425:29
    at /data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/selenium-webdriver/lib/goog/base.js:1582:15
    at [object Object].webdriver.promise.ControlFlow.runInNewFrame_ (/data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver/promise.js:1654:20)
    at notify (/data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver/promise.js:465:12)
    at [object Object].then (/data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver/promise.js:522:7)
    at Object.webdriver.promise.asap (/data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver/promise.js:767:11)
    at newFrame.onComplete (/data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver/promise.js:1670:25)
    at [object Object].webdriver.promise.Frame_.notify_ (/data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver/promise.js:1914:5)
    at [object Object]._onTimeout (/data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver/promise.js:1901:11)
==== async task ====
    at [object Object].webdriver.promise.ControlFlow.wait (/data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver/promise.js:1410:15)
    at [object Object].webdriver.WebDriver.wait (/data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver/webdriver.js:646:21)
    at login (/data/src/github.com/openshift/origin/assets/test/e2e/test.js:38:12)
    at [object Object].<anonymous> (/data/src/github.com/openshift/origin/assets/test/e2e/test.js:209:7)
    at /data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/jasminewd/index.js:94:14
    at [object Object].webdriver.promise.ControlFlow.runInNewFrame_ (/data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver/promise.js:1654:20)
    at [object Object].webdriver.promise.ControlFlow.runEventLoop_ (/data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/selenium-webdriver/lib/webdriver/promise.js:1518:8)
    at [object Object].wrapper [as _onTimeout] (timers.js:261:14)
==== async task ====
Asynchronous test function: beforeEach()
    at [object Object].<anonymous> (/data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/jasminewd/index.js:93:33)
    at [object Object].<anonymous> (/data/src/github.com/openshift/origin/assets/node_modules/protractor/node_modules/minijasminenode/lib/async-callback.js:45:37)
Error
    at [object Object].<anonymous> (/data/src/github.com/openshift/origin/assets/test/e2e/test.js:207:5)
    at [object Object].<anonymous> (/data/src/github.com/openshift/origin/assets/test/e2e/test.js:206:3)
    at Object.<anonymous> (/data/src/github.com/openshift/origin/assets/test/e2e/test.js:3:1)

```

**Comment by jwforres** [comment]
I do now https://github.com/openshift/origin/issues/5881

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/util/clientcache.go [hunk 2]
```diff
@@ -68,6 +68,11 @@ func (c *ClientCache) ClientConfigForVersion(version string) (*client.Config, er
 	client.SetKubernetesDefaults(&config)
 	c.configs[version] = &config
 
+	// `version` does not necessarily equal `config.Version`.  However, we know that we call this method again with
+	// `config.Version`, we should get the the config we've just built.
+	config = config
```
Comment (kargakis) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/util/clientcache.go]: shouldn't this also be a copy?

Comment (deads2k) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/util/clientcache.go]: > shouldn't this also be a copy?

They don't have govet turned on.  It does the same thing.

Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/util/clientcache.go]: that's unfortunate... I like the copy for clarity

Comment (liggitt) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/util/clientcache.go]: without a comment, that line will probably get deleted when they turn govet on

**Pull request merged**